### ISSUE TITLE
[tune/Autoscaler] Autoscaler made demands on a node while knowing this node has died / Cluster hangs randomly when a spot instance worker node is killed.

### ISSUE THREAD

**Issue by YuhangSong** [ray-project/ray#32579]
# What happened + What you expected to happen

1, I run a ray cluster with 80 worker nodes
2, The worker nodes are spot instances, so they get killed randomly
3, Most of the time, it is fine, after some spot instances are killed, some runs are terminated by will be scheduled on other instances. 
4, I also tried manually killing some instances on the aws terminal and the ray cluster runs fine: some runs are terminated and will be scheduled on other instances.
4, But after randomly 5-8 hours, all jobs are hung: you will see error messages like

`(autoscaler +4h28m31s) Error: No available node types can fulfill resource request {'node:172.31.29.192': 0.01}. Add suitable node types to this cluster to resolve this issue.`

Note that this bug is difficult to reproduce because I suspect what happens is
1, most of the time when a spot instance is killed, it is fine, the ray cluster will schedule the trials on that node to other nodes
2, if the spot instance is killed at a very specific moment, maybe at the moment when the ray cluster knows there is a node with ip 172.31.29.192, so request the resource of {'node:172.31.29.192': 0.01}, but right after ray creates the request, the node gets killed. This specific moment might happen very rarely, so most of the time terminating spot instances is fine (even when I manually terminate spot instances), but you will hit such a situation in some hours (in my case normally after 5-8 hours, with >3000 trials completed).

Note that I also referred to #28468, and also update to ray 3.0.0.dev0, it does not resolve this issue, that's why I have the above guess.

@krfricke might have particular insight into this problem.

Then the ray cluster hangs, all jobs hung, and the job keeps logging only the above line, GPU memory is still taken for all GPUs, but CPU or GPU usage is zero.

A follow-up on this to provide more information. I record the screen for 8 hours and found the following behaviour that should indicate a bug. So, in the following moment, the node with ip `172.31.22.221` is identified as dead. For now everything is fine, the demands are made to other alive nodes.

![image](https://user-images.githubusercontent.com/28839288/219306964-4e6f951b-6e30-41ae-a83c-12ed430cf480.png)

However, in the next few seconds, a demand is made to run on this node! See the following:

![image](https://user-images.githubusercontent.com/28839288/219307492-c7ff079c-309e-4663-92fe-eff27eb06d60.png)

In other words, the autoscaler knows a node has died but still schedules a job on it (making resource demands on it).

### Versions / Dependencies

The issue is observed with ray 3.0.0.dev0 and ray 2.2.0

### Reproduction script

It is not really possible to reproduce this issue deterministically. See my above description.

### Issue Severity

High: It blocks me from completing my task.

**Comment by krfricke** [comment]
Thanks for the detailed write-up!

For context, it looks to me like these are actors used for syncing data between nodes (internally these are called `_PackActor`.

To work around the issue, you could pass an `SyncConfig(upload_dir="s3://your-bucket/directory")` to synchronize files using S3 - it will not use the packing actors anymore.

After https://github.com/ray-project/ray/pull/32016 we should use NodeAffinityScheduling and not the `0.01` resources for scheduling anymore. Can you check `ray.__commit__` to see which commit you're running on?

I think a main gap here is that the custom resources scheduling will just wait forever, as it's just a resource, and Ray won't fail the task being started as resources could come up again. I hope the `NodeAffinityScheduling` will react differently.

Another bug seems to be that we apparently don't time out these synchronizations. We should fix this.

Lastly, we don't have extensive tests for the synchronization on node failure case. Note that we mostly optimize for external storage (e.g. S3) for production use cases. Still, we should improve our testing here.

Please do let me know if you can work around this by using S3. The Ray commit hash would also be helpful to further look into this issue.

**Comment by YuhangSong** [comment]
Hi Kai,

Thanks a lot for the quick reply! The ray.__commit__ gives '6708b31fbd7ced3039005e16d62af5c13069efef'. Does that inform anything about this bug?

I am testing your suggested workarounds and will keep you updated soon!

**Comment by YuhangSong** [comment]
Hi Kai,

Using external storage seems to have solved this problem (at least so far).

Updating to more recent commits in the master branch also solves the problem but there are various other bugs.

Thanks a lot for the help (I would not be able to imagine that the problem is actors for packing and syncing results myself!).
